src/ngx_http_lua_worker_thread.c

```diff
@@ -0,0 +1,437 @@
+
+    ctx->rc = lua_pcall(vm, ctx->n_args, LUA_MULTRET, 0);
+}
+
+
+/* executed in nginx event loop */
+static void
+ngx_http_lua_my_thread_completion(ngx_event_t *ev)
+{
+    ngx_http_lua_my_thread_ctx_t     *myctx;
+    int                               is_abort;
+    lua_State                        *L;
+
+    ngx_http_request_t      *r;
+    ngx_connection_t        *c;
+
+    int              nresults;
+    size_t           len;
+    const char      *str;
+    int              i;
+
+    int              rc;
+
+    ngx_http_lua_ctx_t *ctx;
+
+    lua_State *vm;
+
+    myctx = ev->data;
+
+    ngx_thread_mutex_lock(&myctx->ctx->mutex, ngx_cycle->log);
+    is_abort = myctx->is_abort;
+    ngx_thread_mutex_unlock(&myctx->ctx->mutex, ngx_cycle->log);
+
+    if (is_abort) {
+        ngx_http_lua_put_task_ctx(myctx->ctx);
+        return;
+    }
+
+    L = myctx->wait_co_ctx->co;
+
+    r = ngx_http_lua_get_req(L);
+
+    if (r == NULL) {
+        ngx_http_lua_put_task_ctx(myctx->ctx);
+        return;
+    }
+
+    c = r->connection;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+
+    if (ctx == NULL) {
+        ngx_http_lua_put_task_ctx(myctx->ctx);
+        return;
+    }
+
+    vm = myctx->ctx->vm;
+
+    if (myctx->rc != 0) {
+        str = lua_tolstring(vm, 1, &len);
+        lua_pushboolean(L, 0);
+        lua_pushlstring(L, str, len);
+        nresults = 2;
+
+    } else {
+        /* copying return values */
+        lua_pushboolean(L, 1);
+        nresults = lua_gettop(vm);
+        for (i = 1; i <= nresults; i++) {
+            ngx_http_lua_xcopy(vm, L, i, 1);
+        }
+        nresults += 1;
+    }
+
+    ngx_http_lua_put_task_ctx(myctx->ctx);
+
+    /* resume the caller coroutine */
+
+    myctx->wait_co_ctx->cleanup = myctx->cleanup;
+    myctx->wait_co_ctx->data = myctx->cleanup_data;
+    ctx->cur_co_ctx = myctx->wait_co_ctx;
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+
+    rc = ngx_http_lua_run_thread(vm, r, ctx, nresults);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                   "lua run thread returned %d", rc);
+
+    if (rc == NGX_AGAIN) {
+        ngx_http_lua_run_posted_threads(c, vm, r, ctx, c->requests);
+        return;
+    }
+
+    if (rc == NGX_DONE) {
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        ngx_http_lua_run_posted_threads(c, vm, r, ctx, c->requests);
+        return;
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return;
+    }
+}
+
+
+static void
+ngx_http_lua_worker_thread_cleanup(void *data)
+{
+    ngx_http_lua_my_thread_ctx_t *ctx = data;
+
+    ngx_thread_mutex_lock(&ctx->ctx->mutex, ngx_cycle->log);
+    ctx->is_abort = 1;
+    ngx_thread_mutex_unlock(&ctx->ctx->mutex, ngx_cycle->log);
+
+    if (ctx->cleanup) {
+        (ctx->cleanup)(ctx->cleanup_data);
+    }
+}
+
+
+static int
+ngx_http_lua_run_worker_thread(lua_State *L)
```

Comment:
we should not use Lua C API for such new APIs, we only accept the pure ffi way now.
please follow the style as ngx.semaphore:
https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md